File: lib/java/src/main/java/io/github/brenbar/uapi/internal/GenerateRandomStruct.java
package io.github.brenbar.uapi.internal;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.github.brenbar.uapi.RandomGenerator;
import io.github.brenbar.uapi.internal.types.UFieldDeclaration;
import io.github.brenbar.uapi.internal.types.UTypeDeclaration;

import static io.github.brenbar.uapi.internal.ConstructRandomStruct.constructRandomStruct;

public class GenerateRandomStruct {
    public static Object generateRandomStruct(Object blueprintValue, boolean useBlueprintValue,
            boolean includeOptionalFields, boolean randomizeOptionalFields, List<UTypeDeclaration> typeParameters,
            List<UTypeDeclaration> generics, RandomGenerator randomGenerator,
            Map<String, UFieldDeclaration> fields) {
        if (useBlueprintValue) {
            final var startingStructValue = (Map<String, Object>) blueprintValue;
            return constructRandomStruct(fields, startingStructValue, includeOptionalFields, randomizeOptionalFields,
                    typeParameters, randomGenerator);
        } else {
            return constructRandomStruct(fields, new HashMap<>(), includeOptionalFields, randomizeOptionalFields,
                    typeParameters, randomGenerator);
        }
    }
}

File: lib/ts/uapi/internal/generation/GenerateRandomStruct.ts
import { RandomGenerator } from 'uapi/RandomGenerator';
import { UFieldDeclaration } from 'uapi/internal/types/UFieldDeclaration';
import { UTypeDeclaration } from 'uapi/internal/types/UTypeDeclaration';
import { constructRandomStruct } from 'uapi/internal/generation/ConstructRandomStruct';

export function generateRandomStruct(
    blueprintValue: any,
    useBlueprintValue: boolean,
    includeOptionalFields: boolean,
    randomizeOptionalFields: boolean,
    typeParameters: UTypeDeclaration[],
    generics: UTypeDeclaration[],
    randomGenerator: RandomGenerator,
    fields: { [key: string]: UFieldDeclaration },
): any {
    if (useBlueprintValue) {
        // Assuming blueprintValue is already a dictionary
        const startingStructValue = blueprintValue as { [key: string]: any };
        return constructRandomStruct(
            fields,
            startingStructValue,
            includeOptionalFields,
            randomizeOptionalFields,
            typeParameters,
            randomGenerator,
        );
    } else {
        return constructRandomStruct(
            fields,
            {},
            includeOptionalFields,
            randomizeOptionalFields,
            typeParameters,
            randomGenerator,
        );
    }
}
